File: CourseApp/RpgSaga/Archer.cs
namespace CourseApp.RpgSaga;

using System;

public class Archer : Hero
{
    public Archer()
        : base()
    {
        Type = "Archer";
    }

    protected override Hero Attack(Hero enemy)
    {
        if (!SpellUsed)
        {
            Console.WriteLine($"{Type} {Name} use spell and attack {enemy.Type} {enemy.Name}, damage = {Damage + 2}, {enemy.Type} {enemy.Name} health = {enemy.Health - Damage}");
            SpellUsed = true;
            enemy.Health -= Damage;
        }
        else
        {
            enemy.Health -= Damage + 2;
            Console.WriteLine($"{this.Type} {this.Name} attack {enemy.Type} {enemy.Name}, damage = {Damage + 2}, {enemy.Type} {enemy.Name} health = {enemy.Health - Damage}");
        }

        return enemy;
    }
}

File: rpgsaga/saga/src/heroes/archer.ts
import { Messenger } from '../messenger';
import { Hero } from './hero';

const ARROWS_DAMAGE = 2;

export class Archer extends Hero {
  heroName: string = 'Лучник';

  constructor(health: number, power: number, name: string, messenger: Messenger) {
    super(health, power, name, messenger);
  }

  attack(enemy: Hero) {
    enemy.getDamage(this.usedAbility ? this.power + ARROWS_DAMAGE : this.power);
    this.sendDamageMessage(enemy);
  }

  useHeroAbility(enemy: Hero) {
    if (!this.usedAbility) {
      enemy.getDamage(ARROWS_DAMAGE);
      this.usedAbility = true;

      this.messenger.sendMessage(
        `(${this.heroName}) ${this.name} использует (Огненные стрелы) на игрока (${enemy.heroName}) ${enemy.name}, он загорается и теряет ${ARROWS_DAMAGE} единицы жизни.`,
      );
    }
  }
}
